maro/simulator/scenarios/data_center/business_engine.py

```diff
@@ -0,0 +1,253 @@
+
+        self._pending_action_vm_id: int = -1
+
+    def _load_configs(self):
+        """Load configurations."""
+
+        with open(os.path.join(self._config_path, "config.yml")) as fp:
+            self._conf = safe_load(fp)
+
+    def _init_machines(self):
+        """Initialize the physical machines based on the config setting. The PM id starts from 0."""
+
+        self._machines: List[PhysicalMachine] = [
+            PhysicalMachine(
+                id=i,
+                cap_cpu=self._conf["pm_cap_cpu"],
+                cap_mem=self._conf["pm_cap_mem"]
+            ) for i in range(self._conf["pm_amount"])
+        ]
+
+    def step(self, tick: int):
+        """Push business to next step.
+
+        Args:
+            tick (int): Current tick to process.
+        """
+        self._tick = tick
+
+        # Update all live VMs memory utilization.
+        self._update_vm_util()
+        # Update all PM memory utilization.
+        self._update_pm_util()
+        # TODO
+        # Generate VM requirement events from data file.
+        # It might be implemented by a for loop to process VMs in each tick.
+        vm_required_evt = self._event_buffer.gen_cascade_event(tick, Events.REQUIREMENTS, payload=None)
+        self._event_buffer.insert_event(vm_required_evt)
+
+    def get_metrics(self) -> dict:
+        """Get current environment metrics information.
+
+        Returns:
+            dict: Metrics information.
+        """
+
+        return DocableDict(
+            metrics_desc,
+            energy_consumption=self._energy_consumption,
+            success_requirements=self._success_requirements,
+            failed_requirements=self._failed_requirements,
+            total_latency=self._total_latency
+        )
+
+    def _register_events(self):
+        # Register our own events and their callback handlers.
+        self._event_buffer.register_event_handler(Events.REQUIREMENTS, self._on_vm_required)
+        self._event_buffer.register_event_handler(Events.FINISHED, self._on_vm_finished)
+
+        # Generate decision event.
+        self._event_buffer.register_event_handler(MaroEvents.TAKE_ACTION, self._on_action_received)
+
+    def _update_vm_util(self):
+        """Update all live VMs memory utilization.
+        The length of VMs utilization series could be difference among all VMs,
+        because index 0 represents the VM's CPU utilization at the tick it starts.
+        """
+
+        for _, vm in self._vm.items():
+            vm.util_cpu = vm.get_util(cur_tick=self._tick)
+
+    def _update_pm_util(self):
+        """Update memory utilization occupied by total VMs on each PM."""
+        for pm in self._machines:
+            total_cpu: int = 0
+            for vm_id in pm.vm_set:
+                vm = self._vm[vm_id]
+                cur_util_cores = vm.util_cpu * vm.req_cpu / 100
+                total_cpu += cur_util_cores
+            pm.util_cpu = total_cpu / pm.cap_cpu * 100
```

Comment:
The unit of all util_cpu should be %, that is, if util_cpu = 20, it means 0.2 actually.
And cur_util_cores means the vm cores which comes from  util_cpu / 100 * req_cpu (bucket count).
That is to say, if a VM req_cpu = 2 (cores), and its current util_cpu is 20 (%), the cur_util_cores is 2 * 20/100 = 0.4
Since all VM's request cpu are different, we have to calculate the sum of total VM cores, and divided by pm_cap_cpu after the summation of VM cores. So, total_cpu/ pm.cap should multiply 100, as we assume it is in the %.
Like, 5 / 32 = 0.15 and I design to store as 15.
Is it better way to store as 0.15 not 15 ?